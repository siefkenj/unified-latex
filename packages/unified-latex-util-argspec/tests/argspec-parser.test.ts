import { describe, it, expect } from "vitest";
import { VFile } from "unified-lint-rule/lib";
import util from "util";
import * as argspecParser from "..";

/* eslint-env jest */

// Make console.log pretty-print by default
const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

function removeWhitespace(x: string) {
    return x.replace(/\s+/g, "");
}

describe("unified-latex-util-argspec", () => {
    let value: string | undefined;
    let file: VFile | undefined;

    const SPEC_STRINGS = [
        "",
        "o m",
        "o m o !o m",
        "!o r() m",
        "O{somedefault} m o",
        "m e^",
        "m e{_^}",
        "s m",
        "v!",
        "d++ D--{def}",
        "O{nested{defaults}}",
        "m ta o o",
        "u{xx;}",
        "u;",
        "u{ }",
        "Ox",
        "r\\abc\\d",
        "R\\a1{default}",
        "D(ab",
        "E{\\token ^}{{D1}2}",
    ];

    for (const spec of SPEC_STRINGS) {
        it(`parses xparse argument specification string "${spec}"`, () => {
            const ast = argspecParser.parse(spec);
            expect(ast).toMatchSnapshot();
            expect(argspecParser.printRaw(ast, true)).toEqual(spec);
        });
    }

    it("Embellishments always return a string", () => {
        let ast = argspecParser.parse("e{{x}y{z}}");
        expect(ast).toEqual([
            { type: "embellishment", tokens: ["x", "y", "z"] },
        ]);
        ast = argspecParser.parse("E{{x}y{z}}{{One}{Two}{Three}}");
        expect(ast).toEqual([
            {
                type: "embellishment",
                tokens: ["x", "y", "z"],
                defaultArgs: ["One", "Two", "Three"],
            },
        ]);
    });
});
