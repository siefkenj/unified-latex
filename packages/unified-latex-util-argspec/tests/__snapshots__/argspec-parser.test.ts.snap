// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unified-latex-util-argspec > parses xparse argument specification string "!o r() m" 1`] = `
[
  {
    "closeBrace": "]",
    "noLeadingWhitespace": true,
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": ")",
    "openBrace": "(",
    "type": "mandatory",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "" 1`] = `[]`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "D(ab" 1`] = `
[
  {
    "closeBrace": "a",
    "defaultArg": "b",
    "openBrace": "(",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "E{\\token ^}{{D1}2}" 1`] = `
[
  {
    "embellishmentDefaultArg": [
      "D1",
      "2",
    ],
    "embellishmentTokens": [
      "\\\\token",
      "^",
    ],
    "type": "embellishment",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "O{nested{defaults}}" 1`] = `
[
  {
    "closeBrace": "]",
    "defaultArg": "nested{defaults}",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "O{somedefault} m o" 1`] = `
[
  {
    "closeBrace": "]",
    "defaultArg": "somedefault",
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "Ox" 1`] = `
[
  {
    "closeBrace": "]",
    "defaultArg": "x",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "R\\a1{default}" 1`] = `
[
  {
    "closeBrace": "1",
    "defaultArg": "default",
    "openBrace": "\\\\a",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "d++ D--{def}" 1`] = `
[
  {
    "closeBrace": "+",
    "openBrace": "+",
    "type": "optional",
  },
  {
    "closeBrace": "-",
    "defaultArg": "def",
    "openBrace": "-",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "m e^" 1`] = `
[
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  {
    "embellishmentTokens": [
      "^",
    ],
    "type": "embellishment",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "m e{_^}" 1`] = `
[
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  {
    "embellishmentTokens": [
      "_",
      "^",
    ],
    "type": "embellishment",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "m ta o o" 1`] = `
[
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  {
    "token": "a",
    "type": "optionalToken",
  },
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "o m o !o m" 1`] = `
[
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "]",
    "noLeadingWhitespace": true,
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "o m" 1`] = `
[
  {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "r\\abc\\d" 1`] = `
[
  {
    "closeBrace": "\\\\d",
    "openBrace": "\\\\abc",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "s m" 1`] = `
[
  {
    "type": "optionalStar",
  },
  {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "u;" 1`] = `
[
  {
    "stopTokens": [
      ";",
    ],
    "type": "until",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "u{ }" 1`] = `
[
  {
    "stopTokens": [
      " ",
    ],
    "type": "until",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "u{xx;}" 1`] = `
[
  {
    "stopTokens": [
      "x",
      "x",
      ";",
    ],
    "type": "until",
  },
]
`;

exports[`unified-latex-util-argspec > parses xparse argument specification string "v!" 1`] = `
[
  {
    "closeBrace": "!",
    "openBrace": "!",
    "type": "verbatim",
  },
]
`;
